### PR TITLE
Fix throwing ambiguity error for convertible unions

### LINKED ISSUE #35855: Getting an ambiguous error when creating a field with a number union type
**Description:**
Even if the number union field has an array with a specific number type, the ambiguous error returns by this function.

**Steps to reproduce:**
``` ballerina
import ballerina/io;

type Movie record {
    string title;
    int[]|decimal[] year;
    string released;
    Person writer;
};

type Person record {
    string fname;
    string lname;
    int age;
};

public function main() returns error? {
    json inceptionJ = {
        title: "Inception",
        year: [2010.0d, 2011.0d],
        released: "16 Jul 2010",
        writer: {
            fname: "Christopher",
            lname: "Nolan",
            age: 30
        }
    };

    Movie cloneWithType = check inceptionJ.cloneWithType(Movie);
    io:println(cloneWithType);
}
```

**Affected Versions:**
Ballerina 2201.0.3

### PR DESCRIPTION
**Pull request by HindujaB**
## Purpose
$subject

Fixes #35855 

## Approach
The issue happens when the type of a complex value is a union and it needed to be converted into another union that contains numeric members. If the target union type contains the exact type of the numeric value it should provide that same typed value even if there are other numeric types in the union members. 
This PR fixes it by validating the type of the value without any conversion before going into a recursive validation. 

## Samples
```ballerina
public function main() returns error? {
    json[] jsonArr = [23.1d];
    string[]|int[]|decimal[] arrayUnionVal = check  jsonArr.cloneWithType(); // result should be decimal[]

    jsonArr = [23.0d, 24];
    int[]|[decimal, int] arrTupleUnionVal = check jsonArr.cloneWithType(); // result should be [decimal, int] 

    jsonArr = [23.0, 24];
    int[]|decimal[]|float[]|error result2 = jsonArr.cloneWithType(); // should throw ambiguity error

    json m = {a: 1, b: 2};
    map<int>|map<decimal> clone = checkpanic m.cloneWithType(); // result should be map<int>
}

```

## Check List 
- [x] Read the [Contributing Guide](https://github.com/ballerina-platform/ballerina-lang/blob/master/CONTRIBUTING.md)
- [ ] Updated Change Log
- [ ] Checked Tooling Support (#<Issue Number>)
- [x] Added necessary tests
  - [x] Unit Tests
  - [ ] Spec Conformance Tests
  - [ ] Integration Tests
  - [ ] Ballerina By Example Tests
- [x] Increased Test Coverage   
- [ ] Added necessary documentation  
  - [ ] API documentation 
  - [ ] Module documentation in Module.md files
  - [ ] Ballerina By Examples

### CODE REVIEW

File: bvm/ballerina-runtime/src/main/java/io/ballerina/runtime/internal/TypeConverter.java [hunk 1]
```diff
@@ -365,6 +365,22 @@ public static Set<Type> getConvertibleTypes(Object inputValue, Type targetType,
         return convertibleTypes;
     }
 
+    private static boolean isDirectlyConvertible(Object inputValue, Type inputValueType, Type memType,
```
Comment (Nadeeshan96) [inline bvm/ballerina-runtime/src/main/java/io/ballerina/runtime/internal/TypeConverter.java:368]: I think we can refactor the same logic not to have duplicate function calls, like done in https://github.com/HindujaB/ballerina-lang/pull/2/files. There I have also added tests with tuples having rest types.

File: bvm/ballerina-runtime/src/main/java/io/ballerina/runtime/internal/TypeConverter.java [hunk 2]
```diff
@@ -268,21 +268,31 @@ public static Set<Type> getConvertibleTypes(Object inputValue, Type targetType,
...
         switch (targetTypeTag) {
             case TypeTags.UNION_TAG:
                 inputValueType = TypeChecker.getType(inputValue);
+
+                Set<Type> directlyConvertibleTypesInUnion = getDirectlyConvertibleTypesInUnion(
+                        inputValue, inputValueType, (BUnionType) targetType);
+                if (directlyConvertibleTypesInUnion.size() == 1) {
+                    return directlyConvertibleTypesInUnion;
+                }
+                if (!directlyConvertibleTypesInUnion.isEmpty()) {
```
Comment (Nadeeshan96) [inline bvm/ballerina-runtime/src/main/java/io/ballerina/runtime/internal/TypeConverter.java:283]: Do we also need to consider the `allowAmbiguity` flag here?

File: bvm/ballerina-runtime/src/main/java/io/ballerina/runtime/internal/TypeConverter.java [hunk 3]
```diff
@@ -268,21 +268,31 @@ public static Set<Type> getConvertibleTypes(Object inputValue, Type targetType,
         int targetTypeTag = targetType.getTag();
         int initialErrorCount = errors.size();
 
+        String errorMessage =
```
Comment (Nadeeshan96) [inline bvm/ballerina-runtime/src/main/java/io/ballerina/runtime/internal/TypeConverter.java:271]: Shall we rename this to something like `ambiguousTargetErrMsg` and remove unnecessary spaces, if it is better

File: bvm/ballerina-runtime/src/main/java/io/ballerina/runtime/internal/TypeConverter.java [hunk 4]
```diff
@@ -365,6 +374,24 @@ public static Set<Type> getConvertibleTypes(Object inputValue, Type targetType,
         return convertibleTypes;
     }
 
+    private static Set<Type> getDirectlyConvertibleTypesInUnion(Object inputValue, Type inputValueType,
+                                                                BUnionType targetType) {
+        boolean isInputValueSimpleBasicType = isSimpleBasicType(inputValueType);
+        Set<Type> directlyConvertibleTypes = new LinkedHashSet<>();
+        for (Type memType : targetType.getMemberTypes()) {
+            if (isInputValueSimpleBasicType) {
+                if ((inputValueType == memType) || isIntegerSubtypeAndConvertible(inputValue, memType)) {
```
Comment (warunalakshitha) [inline bvm/ballerina-runtime/src/main/java/io/ballerina/runtime/internal/TypeConverter.java:383]: Shalle we merge these if and else conditions to one if condition?

**Comment by Nadeeshan96** [comment]
LGTM

**Review by warunalakshitha — APPROVED** [review]
LGTM

**Pull request merged**